### PR TITLE
A thorough overhaul of the config module.

### PR DESCRIPTION
**Pull request by LintaoAmons**
- Remove the Flag file and the init process
- use `vim.g.scratch_config` to simplify the configuration module
- still use json module to allow the addition of ad-hoc file types.

### CODE REVIEW

**Comment by LintaoAmons** [comment]
#35

File: lua/scratch/config.lua [hunk 1]
```diff
@@ -48,207 +56,80 @@ local default_config = {
...
-  -- write the scratch_file_dir into CONFIG_FILE_PATH file
-  local dir_path = vim.fn.fnamemodify(CONFIG_FILE_FLAG_PATH, ":h")
-  if vim.fn.isdirectory(dir_path) == 0 then
-    vim.fn.mkdir(dir_path, "p")
+  local json_config_path = user_config and user_config.json_config_path or default_json_config_path
+  local json_config
+  if json.file_exists(json_config_path) then
+    json_config = json.read_json_file(json_config_path)
+  else
+    json.write_json_file(default_config, json_config_path)
```
Comment (Rizhiy) [inline lua/scratch/config.lua:116]: Why write `default_config`?

If the user wants to use `json_config` they can just save the file themselves.

File: lua/scratch/config.lua [hunk 2]
```diff
@@ -48,207 +56,80 @@ local default_config = {
...
-local function validate_abspath(path)
-  -- Check if path contains any invalid characters
-  local valid_chars = "[^%w/%.%-%_]+"
-  if vim.fn.has("win32") == 1 then
-    valid_chars = "[^%w%.:%\\%-%_]+"
-  end
-  if string.match(path, valid_chars) then
-    return false
+---@return Scratch.Config
+local function get_and_update_config()
```
Comment (Rizhiy) [inline lua/scratch/config.lua:83]: Isn't `vim.g.scratch_config` already set during setup? Why have another function?

Comment (Rizhiy) [reply lua/scratch/config.lua:83]: If you want to change something during runtime, can't you just update it directly?

File: lua/scratch/scratch_file.lua [hunk 3]
```diff
@@ -16,22 +15,22 @@ local function write_lines_to_buffer(lines)
 end
 
 local function hasDefaultContent(ft)
-  local config_data = config.getConfig()
+  local config_data = config.get_and_update_config()
```
Comment (Rizhiy) [inline lua/scratch/scratch_file.lua:18]: I think it makes sense to just use `vim.g.scratch_config` directly. Here and afterwards.

File: lua/scratch/utils.lua [hunk 4]
```diff
@@ -102,4 +102,118 @@ function M.getSelectedText()
...
+---			city = "New York",
+---			country = "USA",
+---			state = "California",
+---		},
+---	}
+---
+---@param tbl1 table
+---@param tbl2 table
+---@return table
+local function merge_tables(tbl1, tbl2)
```
Comment (Rizhiy) [inline lua/scratch/utils.lua:182]: Doesn't [`vim.tbl_deep_extend`](https://neovim.io/doc/user/lua.html#vim.tbl_deep_extend%28%29) do the same?

File: lua/scratch/utils.lua [hunk 5]
```diff
@@ -102,4 +102,118 @@ function M.getSelectedText()
...
 end
 
-return M
+---@param msg string
+local function log_err(msg)
+  vim.notify(msg, vim.log.levels.ERROR, { title = "easy-commands.nvim" })
+end
+
+-- Helper function to check if a table is array-like
+local function is_array_like(tbl)
```
Comment (Rizhiy) [inline lua/scratch/utils.lua:111]: https://neovim.io/doc/user/lua.html#vim.isarray()

**Review by Rizhiy — COMMENTED** [review]
Thanks, I appreciate the changes. Two broad comments:
1. I think the logic to load json_config is a bit convoluted (see comments). Just check if it exists during setup and load if required. No need to write defaults and manually reload every time. If you want to modify during runtime, I think you can just call config again.
2. Seems to me you are re-inventing some standard functions. I would recommend checking existing [vim functionality](https://neovim.io/doc/user/lua.html#lua-vim) and [plenary](https://github.com/nvim-lua/plenary.nvim) first. No need to re-invent the wheel)

**Comment by LintaoAmons** [comment]
Thanks for your review~
It makes sense, I will look into it later

File: lua/scratch/config.lua (hunk 2 above)
Comment (LintaoAmons) [reply lua/scratch/config.lua:83]: I expose an API to allow user to manually update their json config during runtime~
So I hope the config will always have the up-to-date version~

File: lua/scratch/scratch_file.lua (hunk 3 above)
Comment (LintaoAmons) [reply lua/scratch/scratch_file.lua:18]: Yeah, initially I use just vim.g.scratch_config, but then I thought what if user changed their config in an unexpected way, I mean directly to their config json during runtime

File: lua/scratch/utils.lua (hunk 4 above)
Comment (LintaoAmons) [reply lua/scratch/utils.lua:182]: if the table has a array-like field, will it merge the array's element to form a new array or just use one array value from one table?
I'll take a look

File: lua/scratch/config.lua (hunk 1 above)
Comment (Rizhiy) [reply lua/scratch/config.lua:116]: IMHO, if you want to make things easier for the user then having another file is already a poor choice)
It is usually best to keep things that belong together - close together.

I would recommend following the standard approach of how other plugins do it, i.e. just use `require("scratch").setup(opts)`.

For the instructions a good readme is usually better.

File: lua/scratch/config.lua (hunk 2 above)
Comment (Rizhiy) [reply lua/scratch/config.lua:83]: But this is not an API, this function is being used by every other function.

Having a function to read changes from JSON is fine, but then normal functions should just use `vim.g.scratch_config` directly.

File: lua/scratch/scratch_file.lua (hunk 3 above)
Comment (Rizhiy) [reply lua/scratch/scratch_file.lua:18]: Then they should call that function themselves)
If you want to make it easier for the user, then expose it as a vim command. Something like `:ScractchReloadJson`.

**Comment by LintaoAmons** [comment]
> I would say if you want to make things easy for the user, then you should follow standard conventions/approach used by other plugins, since users most likely are already familiar with it.
> 
> AFAIK, lazy is currently the most popular plugin manager for neovim, so making your plugin work best with it is probably the best approach.

Yes, and also it's simpler if I just need to expose one setup function to user. But I never figure out how lazy.nvim can reload the config in runtime. 

Often the case, when I want to have a scratch file, it maybe a brand new type(cause I want a scratch file to try it out). I don't want to config the plugin, then restart the nvim then have that file type available.

But Yes, I think a better solution is just remove the json config, and try to understand how lazy.nvim reload at the runtime to achieve the same behaviour.

**Comment by LintaoAmons** [comment]
@Rizhiy Can have a look~ I add one more option when user choose their filetype. to achieve the same behaviour. create another filetype without change the config and restart neovim

**Comment by Rizhiy** [comment]
Hi, looks fine to me. Can't really comment on advanced usage, since I just use it for python)

One thing I would recommend, is to check for old flag existing and warn user that it is no longer used + give option to delete it. Many people auto-update their plugins without reading the docs and might be surprised if they just stop working without warning.

**Comment by LintaoAmons** [comment]
Thx~ Yes, makes sense. Will do it

**Pull request merged**